modules/blocks/utils.js

```diff
@@ -332,18 +336,119 @@ Utils.prototype.loadBlocksData = function(filter, cb, tx) {
 			params.limit = realLimit;
 			params.height = height;
 
+			const mergedParams = Object.assign({}, filter, params);
+			const queryFilters = {};
+
+			if (!mergedParams.id && !mergedParams.lastId) {
+				queryFilters.height_lt = mergedParams.limit;
+			}
+
+			if (mergedParams.id) {
+				queryFilters.id = mergedParams.id;
+			}
+
+			if (mergedParams.lastId) {
+				queryFilters.height_gt = mergedParams.height;
+				queryFilters.height_lt = mergedParams.limit;
+			}
+
 			// Retrieve blocks from database
-			// FIXME: That SQL query have mess logic, need to be refactored
-			(tx || library.db).blocks
-				.loadBlocksData(Object.assign({}, filter, params))
-				.then(blockRows => setImmediate(cb, null, blockRows));
+			return library.storage.entities.Block.get(
+				queryFilters,
+				{
+					extended: true,
+					limit: null,
+					sort: ['height'],
+				},
+				tx
+			).then(blockRows => {
```

Comment:
I agree this code block is unnecessarily complex but the previous `Queries.loadBlocksData` query was equally complex so I tried to match the query exactly as it was before to avoid mistakes and make it easier to review this PR.

My suggestion is to open another issue to refactor this method making it more clear and readable.